src/loader/bgw_launcher.c

```diff
@@ -660,16 +765,35 @@ ts_bgw_cluster_launcher_main(PG_FUNCTION_ARGS)
 	while (true)
 	{
 		int			wl_rc;
+		long		timeout = -1;
+		bool		handled_message = false;
+		int			wakeEvents = WL_LATCH_SET | WL_POSTMASTER_DEATH;
 
 		CHECK_FOR_INTERRUPTS();
+
 		stopped_db_schedulers_cleanup(db_htab);
-		if (launcher_handle_message(db_htab))
+		handled_message = launcher_handle_message(db_htab, &createdb_entries);
+
+		handle_finished_createdb_entries(db_htab, &createdb_entries);
+
+		/* If we have createdbs to wait on, set a low timeout */
+		if (list_length(createdb_entries) > 0)
+		{
+			timeout = BGW_WAIT_MESSAGE_INTERVAL;
+			wakeEvents |= WL_TIMEOUT;
+		}
+
+		/*
+		 * Ignore the WaitLatch only if we have handled something AND we
```

Comment:
This logic doesn't make sense to me. I think the original logic was to go through the loop without waiting until you ran out of messages to handle. Not sure why we want to go through the waitlatch if there are createdb messages and handled_messages == true. I would move `handle_finished_createdb_entries` above `launcher_handle_message` and keep the old continue logic